update readme (#433)
aquasecurity/tracee: README.md

# Tracee - Container and system tracing using eBPF

> Tracee is a lightweight and easy to use tracing tool for Linux, which is focused on security and forensics.

Reviewer: I think we need to keep some of the intro, don't you think?

Author: I think we either say something general about unique features or don't say anything and let the demo speak for itself. what do you prefer?

Reviewer: I think we need here at least one paragraph which describes tracee

[Check out this quick demo of tracee](https://youtu.be/WTqE2ae257o)

## Getting started

> ### Prerequisites

Author: the thing is that building the ebpf probe IS part of running tracee. So I think it would be misleading if we rename it to "building", as people would think it just prereqs to build the executable, but these are actually required on their target environments

> ### Prerequisites

Reviewer 2: Ah I see, if the object doesn't exist at $TRACEE_BPF_FILE. Yea this is a little tricky to document and implement. We should explore embedding the bpf object file. I think this all gets easier once we have CO:RE support.

This is fine for now 👍

### Quickstart with Docker

```bash
docker run --name tracee --rm --privileged --pid=host -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro -v /tmp/tracee:/tmp/tracee aquasec/tracee:latest
```

> Note: You may need to change the volume mounts for the kernel headers based on your setup. See [Linux Headers](#Linux-Headers) section for info.

This will run Tracee with no arguments, which defaults to collecting a useful default set of events from all processes and print them in a table to standard output.

### Getting Tracee

You can obtain Tracee in any of the following ways:
1. Download from the [GitHub Releases](https://github.com/aquasecurity/tracee/releases) (`tracee.tar.gz`).
2. Use the docker image from Docker Hub: `aquasec/tracee` (includes all the required dependencies).
3. Build the executable from source using `make build`. For that you will need additional development tooling.
4. Build the executable from source in a Docker container which includes all development tooling, using `make build DOCKER=1`.

### Setup options

> Tracee is made of an executable (`tracee`) that drives the eBPF program, and the eBPF program itself (`tracee.bpf.$kernelversion.$traceeversion.o`). When the `tracee` is started, it will look for the eBPF program in specific places and if not found, it will attempt to build the eBPF program automatically before it starts (you can control this using the `--build-policy` flag).

Author: thanks @grantseltzer I've applied some of you're various suggestions into the current commit, instead of accepting the suggestion, because we generally like to keep the commit history simple.

# Tracee - Container and system tracing using eBPF

> Tracee is a lightweight and easy to use container and system tracing tool. It allows you to observe system calls and other system events in real-time. A unique feature of Tracee is it's comprehensive filtering abilities which allows you to focus on relevant events instead of every single thing that happens on the system. Unlike other tracing tools, Tracee is a security tool, which is demonstrated by features like capturing IoC artifacts from running applications, tracing non-syscall security events, and producing security insights.

Author: I wanted to remove this whole uniqueness part, but I remember you wanted to add it in the first place. are you ok with removing it? so it will be just what you wrote, and the rest they can read about later

> Tracee is a lightweight and easy to use container and system tracing tool. It allows you to observe system calls and other system events in real-time. A unique feature of Tracee is it's comprehensive filtering abilities which allows you to focus on relevant events instead of every single thing that happens on the system. Unlike other tracing tools, Tracee is a security tool, which is demonstrated by features like capturing IoC artifacts from running applications, tracing non-syscall security events, and producing security insights.

Reviewer: The situation is now different, because we no longer filter new processes by default, so modifications are needed, and I understand that.
If you are writing about the unique filters as I wrote - that should be enough

Alternatively, running as `root` or with the `--privileged` flag of Docker, is an easy way to start.

#### Linux Headers

In order to compile the eBPF program, Tracee needs some of the Linux kernel headers. Depending on your Linux distribution, there are different ways to obtain it, and different places they end up at.  

> Note that it's important that the kernel headers match the exact version of kernel you are running. To check the current kernel version, run the command `uname -r`

Here are a few examples of how to obtain Linux kernel headers in different Linux distributions:

Distribution | Command | Installation location
--- | --- | ---
Ubuntu | apt-get install linux-headers | /usr/bin

> 

Author: thanks @grantseltzer is the kernel-devel actually required? if you have a box to verify this (as well as the install location), that would be great since we want to be sure before adding to the readme

> 

Author: pinging @simar7 if you have anything to add here

> 

Reviewer: For manjaro it's just /lib/modules/$(uname -r)/build/include/linux without any links

> 

Author: how do you install headers on manjaro?

> 

Reviewer: sudo pacman -S linux-headers

> 

Reviewer 3: > pinging @simar7 if you have anything to add here

Should we add the section on selinux somewhere? Doesn't have to be here but I feel lots of people will trip considering how it's setup (enforcing) out of the box on certain distros (centos)

> When building the eBPF object file using Tracee's `make bpf`, it will assume the headers to be found in a common location: `/lib/modules/${kernel_version}/build`. If you have the headers elsewhere, you need to provide them to `make` using `KERN_HEADERS` make variable.

Author: it's actually a make variable, not an environment variable

> When building the eBPF object file using Tracee's `make bpf`, it will assume the headers to be found in a common location: `/lib/modules/${kernel_version}/build`. If you have the headers elsewhere, you need to provide them to `make` using `KERN_HEADERS` make variable.

Reviewer 2: Ah ok, got it confused with `KERN_SRC`

Alternatively, running as `root` or with the `--privileged` flag of Docker, is an easy way to start.

#### Linux Headers

In order to compile the eBPF program, Tracee needs some of the Linux kernel headers. Depending on your Linux distribution, there are different ways to obtain it, and different places they end up at.  

> Note that it's important that the kernel headers match the exact version of kernel you are running. To check the current kernel version, run the command `uname -r`

> Here are a few examples of how to obtain Linux kernel headers in different Linux distributions:

Reviewer: You write a few, but give only one example

Author: yes, I meant to ask you about manjaro and simar about centos

Author: so how do you do it on manjaro? @yanivagman

Distribution | Command | Installation location
--- | --- | ---
Ubuntu | apt-get install linux-headers | /usr/bin

> When building the eBPF object file using Tracee's `make bpf`, it will assume the headers to be found in common location: `/lib/modules/${kernel_version}/build`. If you have the headers elsewhere, you need to provide them to `make` using `KERN_BLD_PATH` and `KERN_SRC_PATH` for the build and sources appropriately. 

Reviewer: This can be a bit confusing. Not all users are aware of what is build and sources directories (even for us it took some time to figure this out). Also, the more common thing is to have all the headers in one directory, in which case only 'KERN_HEADERS" can be used

7. EVENT - identifies the event (e.g. syscall name)
8. ARGS - list of arguments given to the function

### Customizing the output

Tracee supports different output formats. For example, to choose json output, use `--output json`.

To tell tracee to write events to a file instead of stdout, use `--output out-file:/path/to/file`.

There are different ways you can augment the output to add useful information. For example: `--output eot` will add a terminating event to the stream which is useful if feeding Tracee's output to another program.

For a full list of output options, run `--output help`.

### Selecting what to trace

> Trace output can easily become unwieldily when tracing all of the events from all of the system. Luckily, Tracee has a powerful mechanism to accurately trace just the information that is relevant to you, using the `--trace` flag.

Reviewer 3: ```suggestion
Trace output can easily become unwieldy when tracing all of the events from all of the system. Luckily, Tracee has a powerful mechanism to accurately trace just the information that is relevant to you, using the `--trace` flag.
```

Using the `--trace` you define expressions that tells Tracee what you are interested in by means of event metadata, and process metadata. Only events that match this criteria will be traced.

You can filter by most of the visible fields on a Tracee event. For example to trace only events related to user ID 1000, use `--trace uid=1000`.  
You can combine trace expressions into more complex criteria. For example, to trace only events related to user ID 1000, which come from process ID 1234, use `--trace uid=1000 --trace pid=1234`.  

A special `pid` value is `new` which let's you trace all newly created processes (that were created after Tracee started tracing).  

> Tracee lets you easily trace events that originate in containers using `--trace container` or only new containers (that was created after Tracee started) using `--trace container=new`.

Reviewer 3: ```suggestion
Tracee lets you easily trace events that originate in containers using `--trace container` or only new containers (that were created after Tracee started) using `--trace container=new`.
```

Event metadata can be used in trace expression as well. For example, to trace only `openat` syscalls, use `--trace event:openat`. But you can also filter on a specific argument of the event, e.g `--trace openat.pathname=/bin/ls` which will show only `openat` syscalls that operate on the file `/bin/ls`.

A useful trace mode is the `--trace follow` which, if specified, will trace not only processes that match the given trace expressions, but also their child processes.
For example, the following will trace all the events that originate from zsh shell, including all of the processes that it will spawn: `--trace command=zsh --follow`.

> For a complete list of trace options, run `--trace help`.

Reviewer 3: When I tried to see the help today, I instinctively passed it as the following:
```
./tracee --trace --help
2021/01/22 11:34:22 invalid filter option specified, use '--filter help' for more info
```

I realize we currently only look at the arg to have "help" and not "--help": https://github.com/aquasecurity/tracee/blob/main/main.go#L345-L347

Should we add "--help" as well?

Reviewer 3: Should we return an error (invalid filter) in this case: `./tracee --trace 'uid<0'` same for `./tracee --trace 'pid<0'`? currently we start tracing

Reviewer 3: What is expected to happen with `--trace '!c'`? I currently don't see we handle it here: https://github.com/aquasecurity/tracee/blob/main/main.go#L440 but if `--trace '!container'` and `--trace c` are a thing, intuitively I think `--trace '!c'` would be too.

Reviewer: Doing --trace '!c' works exactly as you expect. Give it a try (and then you can also see why the hasprefix in line 440 works - it's 'f' being the prefix of "!container")

Reviewer 3: Ah yes - I missed that detail. It does get caught by '!c'. I think we should add it to the readme as an example.

Reviewer: Regarding pid/uid<0 - I agree. We should produce an error when such values are given.
Also agree about the --help thing (and -h as well).
These two, of course, are for different commits.

Regarding the '!c' in the readme - it previously was an example in the --trace help, but I removed it as I already put there "container", "c" and "!container", and I thought it was only logical that people will understand what "!c" does.

Reviewer 3: Roger - I created an issue to track it: https://github.com/aquasecurity/tracee/issues/450

### Capturing artifacts

Tracee has a unique feature that lets you capture interesting artifacts from running applications, using the `--capture` flag.

All captured artifacts are saved in Tracee's "output directory" which can be configured using `--capture dir:/path/to/dir`.

> Tracee can capture the following types of IoCs:

Reviewer 3: What is IoC?

Reviewer 2: Indicator of compromise.

Reviewer 3: Roger - we should put that here as it's the first usage within this document.
```suggestion
Tracee can capture the following types of Indicator of Compromise (IoC):
```

You can filter by most of the visible fields on a Tracee event. For example to trace only events related to user ID 1000, use `--trace uid=1000`.  
You can combine trace expressions into more complex criteria. For example, to trace only events related to user ID 1000, which come from process ID 1234, use `--trace uid=1000 --trace pid=1234`.  

A special `pid` value is `new` which let's you trace all newly created processes (that were created after Tracee started tracing).  
Tracee lets you easily trace events that originate in containers using `--trace container` or only new containers (that was created after Tracee started) using `--trace container=new`.

Event metadata can be used in trace expression as well. For example, to trace only `openat` syscalls, use `--trace event:openat`. But you can also filter on a specific argument of the event, e.g `--trace openat.pathname=/bin/ls` which will show only `openat` syscalls that operate on the file `/bin/ls`.

A useful trace mode is the `--trace follow` which, if specified, will trace not only processes that match the given trace expressions, but also their child processes.

> For example, the following will trace all the events that originate from zsh shell, including all of the processes that will spawns: `--trace command=zsh --follow`.

Reviewer: that will spawns -> that it will spawn, or, that is spawns

For a full list of output options, run `--output help`.

### Selecting what to trace

Trace output can easily become unwieldily when tracing all of the events from all of the system. Luckily, Tracee has a powerful mechanism to accurately trace just the information that is relevant to you, using the `--trace` flag.
Using the `--trace` you define expressions that tells Tracee what you are interested in by means of event metadata, and process metadata. Only events that match this criteria will be traced.

You can filter by most of the visible fields on a Tracee event. For example to trace only events related to user ID 1000, use `--trace uid=1000`.  
You can combine trace expressions into more complex criteria. For example, to trace only events related to user ID 1000, which come from process ID 1234, use `--trace uid=1000 --trace pid=1234`.  

A special `pid` value is `new` which let's you trace all newly created processes (that were created after Tracee started tracing).  

> Tracee let's use easily trace events that originate in containers using `--trace container` or only new containers (that was created after Tracee started) using `--trace container=new`.

Reviewer: Tracee let's use? Not sure I understand this sentence

> Event metadata can be used in trace expression as well. For example, to trace only `operat` syscalls, use `--trace event:openat`. But you can also filter on a specific argument of the event, e.g `--trace openat.pathname=/bin/ls` which will show only `openat` syscalls that operate on the file `/bin/ls`.

Reviewer: operat -> openat

> For a complete list of trace options, run `--trace help`.

Reviewer: I think that the "follow" flag is very useful and should be part of the help, don't you think?

### Capturing artifacts

Tracee has a unique feature that lets you capture interesting artifacts from running applications, using the `--capture` flag.

All captured artifacts are saved in Tracee's "output directory" which can be configured using `--capture dir:/path/to/dir`.

Tracee can capture the following types of IoCs:

1. Written files: Anytime a file is being written to, the contents of the file will be captured. This can be filtered to select specific file patterns to capture.
2. Executed files: Anytime a binary is being executed, the binary file will be captured. If the same binary is executed multiple times, it will be captured just once.
3. Memory files: Anytime a "memory unpacker" is detected, the suspicious memory region will be captured. This is triggered when memory protection changes from Write+Execute to Write.

To use, `--capture exec`, `--capture mem`, and `--capture write` capture executed, memory, and written files respectively. 

> Written files can be filtered using an optional path prefix, e.g `--capture write=/etc/*` will capture anything written blow `/etc/`.

Reviewer: If you are writing about the write filter here, I don't think it is necessary in point 1 above
